azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -757,6 +792,117 @@ public LogData getExecutionJobLog(final ExecutableFlow exFlow, final String jobI
     return jobStats;
   }
 
+  @Override
+  public String getJobLinkUrl(final ExecutableFlow exFlow, final String jobId, final int attempt) {
+    if (!this.azkProps.containsKey(ConfigurationKeys.AZKABAN_RM_JOB_LINK) || !this.azkProps
+        .containsKey(ConfigurationKeys.AZKABAN_JHS_JOB_LINK) || !this.azkProps
+        .containsKey(ConfigurationKeys.AZKABAN_SHS_JOB_LINK)) {
+      return null;
+    }
+
+    final String applicationId = getApplicationId(exFlow, jobId, attempt);
+    if (applicationId == null) {
+      return null;
+    }
+
+    final URL url;
+    final String jobLinkUrl;
+    boolean isRMJobLinkValid = true;
+
+    try {
+      url = new URL(this.azkProps.getString(ConfigurationKeys.AZKABAN_RM_JOB_LINK) + applicationId);
+      final HttpURLConnection connection = loginAuthenticatedURL(url);
+
+      try (final BufferedReader in = new BufferedReader(
+          new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
+        String inputLine;
+        while ((inputLine = in.readLine()) != null) {
+          if (FAILED_TO_READ_APPLICATION_PATTERN.matcher(inputLine).find()
```

Comment:
Are you parsing the html content here? 
How reliable is this method? What if newer versions change the content?

Reply:
Yes, I'm parsing the html content. I've discussed it with Min and we couldn't think of any better way to do it. If anything changes in Hadoop side, then yes we have to change our code accordingly.

Reply:
How do you plan to catch compatibility issues before users do when upgrading Hadoop or Spark?

Reply:
We can create a ticket to keep Hadoop team in the loop and notice us if anything changes related to Hadoop/Spark logs. Worst case would be the "Hadoop Job Log" button doesn't show up or the link is invalid. This is not affecting core Azkaban functionality and users still have other ways to find the job logs. This is a nice-to-have feature and has low risk. So I would say we do our best to communicate with Hadoop team to keep us updated with new changes. Let me know if you have any better suggestions.

Reply:
> We can create a ticket to keep Hadoop team in the loop and notice us if anything changes related to Hadoop/Spark logs.

I don't think this is a realistic expectation. You can test by asking them.

> Worst case would be the "Hadoop Job Log" button doesn't show up or the link is invalid. This is not affecting core Azkaban functionality and users still have other ways to find the job logs. This is a nice-to-have feature and has low risk. So I would say we do our best to communicate with Hadoop team to keep us updated with new changes.

Understood. I would agree not to make the investment to make the code robust and have good test coverage now if this is an experimental feature and we are just testing if users will find it useful and this is not the final state. Is this the case? 
Otherwise, it is a slippery slope IMHO. How would you feel about a product that most of the "non-essential" features are not reliable? What are the "non-essential" features anyway? 

> Let me know if you have any better suggestions.
How difficult will it be to write an automated test to verify this dependency/assumption? Such a test only needs to run before any Hadoop/spark upgrade.

Reply:
Of course we would not want a product with unreliable non-essential feature. I might not conveying myself clearly so that there is a misunderstanding here. The difficulty I have to admit is that this feature highly depends on Hadoop side unlike other Azkaban features which we could test before new Azkaban release. Today Azkaban doesn't have UI test framework, otherwise it would be easy to add a test for this hadoop log button. Also the test case is not easy if you want to catch all possible scenarios. For now, I would prefer the most straightforward way which is to just click the button and see if it works when there is any related Hadoop upgrade.

Reply:
I don't think testing this assumption would require testing UIs. The assumption here is not related to how the AZ button is presented, is it?

> Also the test case is not easy if you want to catch all possible scenarios. 
What are the scenarios you have in mind? 

Happy to discuss in person.